### PR TITLE
Migrate data from old vfs paths to new local volumes path.

### PR DESCRIPTION
**Pull request by calavera**
This is a bug upgrading from Docker 1.6 to Docker 1.7.

Before:

dockercompat-before

After:

dockercompat-after

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

File: daemon/volumes.go [hunk 1]
```diff
@@ -287,6 +290,27 @@ func (daemon *Daemon) verifyVolumesInfo(container *Container) error {
 	return container.ToDisk()
 }
 
+func (daemon *Daemon) migrateVolume(id, hostPath string) error {
```
Comment (crosbymichael) [inline daemon/volumes.go]: Should this only happen if we are using the local driver?

Comment (calavera) [inline daemon/volumes.go]: yeah, this is only called when there is a `/var/lib/docker/vfs/dir/VOLUME_ID` directory, which only happens using the local driver in versions < 1.7. After 1.7 all volumes are in `/var/lib/docker/volumes`.

Comment (crosbymichael) [inline daemon/volumes.go]: ok, cool

File: daemon/volumes.go [hunk 2]
```diff
@@ -287,6 +290,27 @@ func (daemon *Daemon) verifyVolumesInfo(container *Container) error {
...
+	volumePath := filepath.Join(daemon.root, defaultVolumesPathName, id)
+	backupPath := filepath.Join(daemon.root, defaultVolumesPathName, id+".back")
+
+	var err error
+	if err = os.Rename(volumePath, backupPath); err != nil {
```
Comment (crosbymichael) [inline daemon/volumes.go]: could you maybe add a comment to this flow, i'm trying to follow but having all the vars named `*Path` make it hard to folow.

Comment (calavera) [inline daemon/volumes.go]: yes

Comment (calavera) [inline daemon/volumes.go]: @crosbymichael I've renamed the variables and add some comments.

File: daemon/volumes.go [hunk 3]
```diff
@@ -287,6 +290,37 @@ func (daemon *Daemon) verifyVolumesInfo(container *Container) error {
...
+			os.Rename(backup, volumeInfo)
+		}
+	}()
+
+	if err = os.Rename(vfs, volumeInfo); err != nil {
+		return err
+	}
+
+	if err = os.RemoveAll(backup); err != nil {
+		return err
```
Comment (icecrime) [inline daemon/volumes.go]: That's a tricky code path: if you fail to remove, you'll try rename `backup` back to `volumeInfo`, which is now taken  (L313). Maybe that's the one error path where we should only log the error?

Comment (calavera) [inline daemon/volumes.go]: agreed, and fixed.

**Comment by icecrime** [comment]
LGTM

**Comment by crosbymichael** [comment]
LGTM

**Comment by jfrazelle** [comment]
cherry-picked

**Pull request merged**